### ISSUE TITLE
Alert Emails to Receive - list keeps populating every time I try to edit the alert

### ISSUE THREAD

**Issue by rbeckman-nextgen** [nextgenhealthcare/connect#1252]
Version: mirth-1.8.2.4358
DB: MySQL
Java: 1.6.0_16

Steps to reproduce:

1. Create an alert and put an e-mail address in the Emails to Receive
2. Go back to the alert and edit

Expected behavior:

Edit happens and the settings take.

What actually happens:

The changes to the everything else seems to work, but the the list Emails to Receive seems to grow automatically. I have one e-mail address in there and every time I go to edit the alert Mirth Connect keeps adding the same e-mail in.

Imported Issue. Original Details:
Jira Issue Key: MIRTH-1252
Reporter: malcomm
Created: 2009-11-13T13:49:51.000-0800

**Comment by rbeckman-nextgen** [comment]
I'm attaching the exported alert: alert.xml.

Imported Comment. Original Details:
Author: malcomm
Created: 2009-11-13T13:52:11.000-0800

**Comment by rbeckman-nextgen** [comment]
Actually ... I just created a brand new alert and it might not anything to do w/ editing the alert. Instead, Mirth might be adding the e-mail addresses whenever an ERROR occurs. Difficult to diagnose, but I know what the symptoms are. Any help?

Imported Comment. Original Details:
Author: malcomm
Created: 2009-11-13T14:37:28.000-0800

**Comment by rbeckman-nextgen** [comment]
I have not been able to reproduce this.  Do you have a sure way of reproducing it with a new alert?

Imported Comment. Original Details:
Author: jacobb
Created: 2009-11-18T15:39:08.000-0800

**Comment by rbeckman-nextgen** [comment]
Hi Jacob - this is still on my radar, but other high priority projects keep getting in the way. I'll hopefully have time in the next couple of days to get this going.

Thank you for the ping -- Marcus

Imported Comment. Original Details:
Author: malcomm
Created: 2009-12-01T16:16:39.000-0800

**Comment by rbeckman-nextgen** [comment]
Any update to this issue?

Imported Comment. Original Details:
Author: jacobb
Created: 2009-12-09T13:41:28.000-0800

**Comment by rbeckman-nextgen** [comment]
This is just not an issue with 1.8.2 version.. its there in 1.8.0 also. We were using MySql as DB, OS - Linux, Java 1.5. I am not sure if this worked in 1.8.1.

Imported Comment. Original Details:
Author: srinivasay
Created: 2009-12-09T22:18:24.000-0800

**Comment by rbeckman-nextgen** [comment]
srinivasa y, what exact behavior do you experience and how do you reproduce it?

Imported Comment. Original Details:
Author: jacobb
Created: 2009-12-11T13:59:08.000-0800

**Comment by rbeckman-nextgen** [comment]
As such I don't have exact steps to reproduce it. We have around 25 channels deployed and couple of alerts. This is what I noticed, every time channels are redeployed its adding duplicate email id's to each alert. Though this didn't happen initially but after some time with regular redeploys and without restarting mirth it all started. I also noticed that when these alerts are disabled, duplicating of email id's also stopped.

Imported Comment. Original Details:
Author: srinivasay
Created: 2009-12-14T22:50:39.000-0800

**Comment by rbeckman-nextgen** [comment]
Still have not been able to reproduce this.  If anyone can give steps for reproducing it I will be able to assign it to a version for fixing the issue.

Imported Comment. Original Details:
Author: jacobb
Created: 2009-12-22T15:45:53.000-0800

**Comment by rbeckman-nextgen** [comment]
Sorry this took so long. I can reproduce this every time by following these steps:

1) Enable an alert for a singe channel
2) Make the channel error out (I'm doing this by doing a test ftp sender and putting a bad host name)

When this happens, the Email Address list for the alert keeps on growing. That is, I will see my e-mail in there only once to start off with, but as soon as I make the channel error out it keeps appending to the list. I'll try and look at the code to see if I can figure out what is going on and debug that way.

Oh and I'm running 1.8.2 (released version).

Thank you,

Marcus



Imported Comment. Original Details:
Author: malcomm
Created: 2010-02-01T20:17:05.000-0800

**Comment by rbeckman-nextgen** [comment]
So I just built Mirth from Eclipse, both 1.8.2 and trunk (2.0.0) and I can't reproduce this with the default derby database. After hooking up MySQL, I can reproduce this every time. Doing some investigation to see why, but thought I'd put that bit of info into the issue.

--
Marcus

Imported Comment. Original Details:
Author: malcomm
Created: 2010-02-04T21:43:10.000-0800

**Comment by rbeckman-nextgen** [comment]
Small correction: I was able to reproduce this only once with a channel that error'd out, subsequent errors did not create additional e-mail addresses. So that was a little unexpected. Next I tried to delete the 2nd e-mail addresses and instead of deleting the address the act of trying to remove the e-mail added one instead of removing one. Each attempt to delete will add e-mail addresses.

Imported Comment. Original Details:
Author: malcomm
Created: 2010-02-04T21:51:52.000-0800

**Comment by rbeckman-nextgen** [comment]
I believe I might have a "fix" for this, so I noticed that the ALERT_EMAIL table was not getting deleted when this call is made:

SqlConfig.getSqlMapClient().delete("Alert.deleteAlert", alert);

I saw this:

CREATE TABLE ALERT_EMAIL
	(ALERT_ID VARCHAR(255) NOT NULL,
	EMAIL VARCHAR(255) NOT NULL,
	CONSTRAINT ALERT_ID_AE_FK FOREIGN KEY(ALERT_ID) REFERENCES ALERT(ID) ON DELETE CASCADE);

The only problem: my tables were all built using MyISAM and not InnoDB. From what I remember and what I've been reading MyISAM does not support FK cascade deletes. To fix this problem I changed the above create statement to be:

CREATE TABLE ALERT_EMAIL_BAR
	(ALERT_ID VARCHAR(255) NOT NULL,
	EMAIL VARCHAR(255) NOT NULL,
	CONSTRAINT ALERT_ID_AE_FK FOREIGN KEY(ALERT_ID) REFERENCES ALERT(ID) ON DELETE CASCADE) ENGINE=InnoDB;

That fixed the problem ... I think I would recommend forcing InnoDB, I'll attach a patch file just in case that's how you want to go.

Imported Comment. Original Details:
Author: malcomm
Created: 2010-02-04T22:52:30.000-0800

**Comment by rbeckman-nextgen** [comment]
Attaching mysql-database.sql.patch which sets ENGINE=InnoDB on all create tables. I did a basic test with a simple file transfer channel. Naturally InnoDB might not be supported by all systems, so this might not be the best, but at least this is a good workaround for those that can use InnoDB.

--
Marcus

Imported Comment. Original Details:
Author: malcomm
Created: 2010-02-04T23:10:02.000-0800

**Comment by rbeckman-nextgen** [comment]
srinivasa y, are you using MyISAM on mysql as well?

Imported Comment. Original Details:
Author: jacobb
Created: 2010-02-05T14:01:27.000-0800

**Comment by rbeckman-nextgen** [comment]
Reproduced this with mysql using MyISAM.  Will remove ALERT_EMAILS table and store csv addresses as a column.

Imported Comment. Original Details:
Author: jacobb
Created: 2010-02-05T15:53:56.000-0800

**Comment by rbeckman-nextgen** [comment]
Hi Jacob,

I am seeing the "ON DELETE CASCADE" on the following tables:

MESSAGE
CHANNEL_STATISTICS
CHANNEL_ALERT

Not sure if this will have impact there as well.

Imported Comment. Original Details:
Author: malcomm
Created: 2010-02-05T16:20:14.000-0800

**Comment by rbeckman-nextgen** [comment]
Changed the mysql create script to create every table with ENGINE=InnoDB.  Did not fix just the alert email table because other problems not exposed in the UI are still caused by using MyISAM.

Imported Comment. Original Details:
Author: jacobb
Created: 2010-02-05T16:34:07.000-0800

**Comment by rbeckman-nextgen** [comment]
I noticed other problems on those tables as well Marcus.  srinivasa y, if any problems still exist with alert emails after changing to InnoDB from MyISAM or you experienced a different issue (different engine or database), please create a new issue.

Imported Comment. Original Details:
Author: jacobb
Created: 2010-02-05T16:35:48.000-0800
